tutorial/practicals/nemo/4_nemo_openacc/README.md

```diff
@@ -0,0 +1,416 @@
+            tloop = node
+            break
+    ACC_KERNELS_TRANS.apply(tloop.loop_body)
+   ```
+
+3. We can then refine our first attempt so that we only collapse latitude
+   loops that are *within* this region (i.e. children of the 'iteration'
+   loop):
+   ```python
+    loops = tnode.walk(Loop)
+    for loop in loops if loop.loop_type == "lat":
+        ACC_LOOP_TRANS.apply(loop)
+   ```
+   At this point, PSyclone should successfully generate valid Fortran
+   with OpenACC directives. However, if we examine the Fortran produced
+   we will see that the LOOP directives currently only specify
+   `INDEPENDENT`, e.g.:
+   ```fortran
+      DO jk = 1, jpk - 1
+        !$ACC LOOP INDEPENDENT
+        DO jj = 2, jpj - 1
+          DO ji = 2, jpi - 1
+   ```
+
+4. How do we add the `COLLAPSE` clause? If we look at the documentation
+   for `ACCLoopTrans` in the
+   [Transformations](https://psyclone.readthedocs.io/en/stable/transformations.html?highlight=accdatatrans#transformations)
+   section of the User Guide, we see that it takes an `options`
+   dictionary argument. We can therefore specify that we want
+   `COLLAPSE(2)` by doing:
+   ```python
+   ACC_LOOP_TRANS.apply(loop, options={"collapse": 2})
+   ```
+
+   The resulting Fortran should now look like:
+   ```fortran
+      DO jk = 1, jpk - 1
+        !$ACC LOOP INDEPENDENT COLLAPSE(2)
+        DO jj = 2, jpj - 1
+          DO ji = 2, jpi - 1
+   ```
+   This option has been found to improve the performance of the NEMO model
+   on GPU by a few percent.
+
+If time allows, you might wish to try modifying the mini-app so that
+at least one of the latitude loops does *not* correspond to a
+tightly-nested loop, e.g.:
+
+```fortran
+    DO jk = 1, jpk - 1
+      DO jj = 2, jpj - 1
+        DO ji = 2, jpi - 1
+	  ...
+	END DO
+	DO ji = 2, jpi - 1
+	  some-new-statements-here
+	END DO
+      END DO
+    END DO
+```
+
+You will then need to generalise your script so that it only applies
+the `ACCLoopTrans` to valid loop nests. (Hint: you will need to examine
+the nodes in the `loop_body` of the candidate latitude loop.)
+
+## Managed Memory ##
+
+In practice, the work being done to extend PSyclone to process the
+whole of the NEMO code is currently using NVIDIA's 'managed memory'
+support. No explicit data regions are added to the code. Instead, the
+run-time system moves data to/from the GPU automatically when page
+faults occur. This was originally intended as being a quick way to get
+something working on the GPU but it has actually proved to work well
+in general.
+
+
+## Conclusion ##
+
+Congratulations! You have now completed the OpenACC part of the
+tutorial.  You should now understand the basic OpenACC transformations
+provided by PSyclone and how these can be used to accelerate
+NEMO-style code on a GPU.
+
+[ref_kernelstrans]: https://psyclone-ref.readthedocs.io/en/latest/_static/html/classpsyclone_1_1transformations_1_1ACCKernelsTrans.html "ACCKernelsTrans"
```

Comment:
Why is this and the datatrans reference here?

Reply:
I think I was having trouble getting the formatting of the links right but I know how to do it now so have put them in-place.